src/xenia/base/cvar.h

```diff
@@ -0,0 +1,258 @@
+  return "\"" + val + "\"";
+}
+
+template <class T>
+std::string CommandVar<T>::ToString(T val) {
+  return std::to_string(val);
+}
+
+template <class T>
+void CommandVar<T>::SetValue(T val) {
+  *currentValue_ = val;
+}
+template <class T>
+std::string ConfigVar<T>::GetCategory() {
+  return category_;
+}
+template <class T>
+std::string ConfigVar<T>::GetConfigValue() {
+  if (this->configValue_) return this->ToString(*this->configValue_);
+  return this->ToString(this->defaultValue_);
+}
+template <class T>
+void CommandVar<T>::SetCommandLineValue(const T val) {
+  this->commandLineValue_ = std::make_unique<T>(val);
+  this->UpdateValue();
+}
+template <class T>
+void ConfigVar<T>::SetConfigValue(T val) {
+  this->configValue_ = std::make_unique<T>(val);
+  this->UpdateValue();
+}
+template <class T>
+void ConfigVar<T>::SetGameConfigValue(T val) {
+  this->gameConfigValue_ = std::make_unique<T>(val);
+  this->UpdateValue();
+}
+
+extern std::map<std::string, ICommandVar*>* CmdVars;
+extern std::map<std::string, IConfigVar*>* ConfigVars;
+
+inline void AddConfigVar(IConfigVar* cv) {
+  if (!ConfigVars) ConfigVars = new std::map<std::string, IConfigVar*>();
+  ConfigVars->insert(std::pair<std::string, IConfigVar*>(cv->GetName(), cv));
+}
+inline void AddCommandVar(ICommandVar* cv) {
+  if (!CmdVars) CmdVars = new std::map<std::string, ICommandVar*>();
+  CmdVars->insert(std::pair<std::string, ICommandVar*>(cv->GetName(), cv));
+}
+void ParseLaunchArguments(int argc, char** argv);
+
+template <typename T>
+T* define_configvar(const char* name, T* defaultValue, const char* description,
+                    const char* category) {
+  IConfigVar* cfgVar =
+      new ConfigVar<T>(name, defaultValue, description, category);
+  AddConfigVar(cfgVar);
+  return defaultValue;
+}
+
+template <typename T>
+T* define_cmdvar(const char* name, T* defaultValue, const char* description) {
+  ICommandVar* cmdVar = new CommandVar<T>(name, defaultValue, description);
+  AddCommandVar(cmdVar);
+  return defaultValue;
+}
+#define DEFINE_double(name, defaultValue, description, category) \
+  DEFINE_CVar(name, defaultValue, description, category, double)
+
+#define DEFINE_int32(name, defaultValue, description, category) \
+  DEFINE_CVar(name, defaultValue, description, category, int32_t)
+
+#define DEFINE_uint64(name, defaultValue, description, category) \
+  DEFINE_CVar(name, defaultValue, description, category, uint64_t)
+
+#define DEFINE_string(name, defaultValue, description, category) \
+  DEFINE_CVar(name, defaultValue, description, category, std::string)
+
+#define DEFINE_bool(name, defaultValue, description, category) \
+  DEFINE_CVar(name, defaultValue, description, category, bool)
+
+#define DEFINE_CVar(name, defaultValue, description, category, type) \
+  namespace cvar {                                                   \
+  type var_##name = defaultValue;                                    \
```

Comment:
Pick one:
1. `type VAR_##name = defaultValue;`, `using cvar::VAR_##name`. `VAR_name` for access.
2. `type name = defaultValue;`, `cvar::name` for access.

Upper case prefix makes things stand out clearly if we're not going to use the namespace. Personally, I'd very much prefer the latter, but either is acceptable.

If you go with the latter, probably better to stuff them into `cvars` namespace instead of `cvar` to make things distinct.

Reply:
Good suggestion, went with the latter approach. All cvars are now in cvars::name